src/common/utils.ts

```diff
@@ -477,3 +483,52 @@ export const objectCheckOneWay = (obj1: any, obj2: any): boolean => {
   }
   return true
 }
+
+/**
+ * Adds the API key to an RPC server url
+ */
+export const addRpcApiKey = (
+  initOptions: EthereumInitOptions,
+  pluginId: string,
+  url: string
+): string => {
+  const regex = /{{(.*?)}}/g
+  const match = regex.exec(url)
+  if (match != null) {
+    const key = match[1]
+    const cleanKey = asEthereumInitKeys(key)
+    const apiKey = initOptions[cleanKey]
+    if (typeof apiKey === 'string') {
+      url = url.replace(match[0], apiKey)
+    } else if (apiKey == null) {
+      throw new Error(`Missing ${cleanKey} in 'initOptions' for ${pluginId}`)
+    } else {
+      throw new Error('Incorrect apikey type for RPC')
+    }
+  }
+  return url
+}
+
+/**
+ * Calls `func` on ethers JsonRpcProviders initialized with configured
+ * RPC servers. Randomizes order priority to distribute load.
+ */
+export const multicastEthProviders = async (props: {
+  func: (ethProvider: ethers.providers.JsonRpcProvider) => Promise<any>
+  rpcServers: string[]
+  initOptions: EthereumInitOptions
+  pluginId: string
+  chainId?: number
+}): Promise<any> => {
+  const { func, rpcServers, initOptions, pluginId, chainId } = props
+  const funcs: Array<() => Promise<any>> = rpcServers.map(
+    rpcServer => async () => {
+      const ethProvider = new ethers.providers.JsonRpcProvider(
+        addRpcApiKey(initOptions, pluginId, rpcServer),
```

Comment:
Add the API key before calling `multicastEthProviders` and now `multicastEthProviders` doesn't need the extra parameters.

Also, instead of `rpcServers: string[]` you can paramterize `providers: ethers.providers.JsonRpcProvider[]`. In the end the function signature could look like:

```ts
export const multicastEthProviders = async <R, P extends ethers.providers.Provider>(props: {
  func: (ethProvider: P) => Promise<R>
  providers: P[]
}): Promise<R>
```

If you want to be more narrow then `Providers` type then change the subtype for `P` to `ethers.providers.JsonRpcProvider`.